### ISSUE TITLE
Replication errors: STATE column is empty

### ISSUE THREAD

**Issue by jh23453** [peterpakos/ipa_check_consistency#34]
I get the following output:
```
FreeIPA servers:    freeipa1    freeipa2    STATE
=================================================
...
Anonymous BIND      YES         YES         OK   
Replication Status  freeipa2 18 freeipa1 0       
=================================================
```
I'd expect OK or ERROR in the state column.
This trace might help to fix:
```
+ local server= agreements= agreement= state=
+ local -i i=0 max=0
+ local dir=/tmp/user/1004/tmp.vDKeVHg4LH/replication
++ sed -e 's@=@\\3D@g' -e 's@,@\\,@g'
+ local 'suffix=dc\3Djochen\,dc\3Dorg'
+ mkdir -p /tmp/user/1004/tmp.vDKeVHg4LH/replication
+ for server in "${SERVERS[@]}"
+ for server in "${SERVERS[@]}"
+ wait
++ awk '{if ($0 ~ /nsDS5ReplicaHost/) printf("%s",$2); if ($0 ~ /nsds5replicaLastUpdateStatus/) printf(" %s\n",$2)}'
++ sed -e 's/\.example.org//g'
++ query_ldap 'cn=replica,cn=dc\3Dexample\,dc\3Dorg,cn=mapping tree,cn=config' nsDS5ReplicaHost nsds5replicaLastUpdateStatus one
++ sed -e 's/Error (//' -e 's/)//'
++ local 'base=cn=replica,cn=dc\3Dexample\,dc\3Dorg,cn=mapping tree,cn=config'
++ local filter=nsDS5ReplicaHost                                                                                                                                        
++ local attr=nsds5replicaLastUpdateStatus                                                                                                                              
++ local scope=one                                                                                                                                                      
++ ldapsearch -o ldif-wrap=no -ZZ -LLLx -h freeipa1.example.org -D 'cn=Directory Manager' -w very-top-secret -s one -b 'cn=replica,cn=dc\3Dexample\,dc\3Dorg,cn=mapping tree,cn=config' nsDS5ReplicaHost nsds5replicaLastUpdateStatus                                                                                                             
++ query_ldap 'cn=replica,cn=dc\3Dexample\,dc\3Dorg,cn=mapping tree,cn=config' nsDS5ReplicaHost nsds5replicaLastUpdateStatus one
++ sed -e 's/Error (//' -e 's/)//'
++ local 'base=cn=replica,cn=dc\3Dexample\,dc\3Dorg,cn=mapping tree,cn=config'
++ sed -e 's/\.example.org//g'
++ local filter=nsDS5ReplicaHost
++ local attr=nsds5replicaLastUpdateStatus
++ local scope=one
++ ldapsearch -o ldif-wrap=no -ZZ -LLLx -h freeipa2.example.org -D 'cn=Directory Manager' -w very-top-secret -s one -b 'cn=replica,cn=dc\3Dexample\,dc\3Dorg,cn=mapping tree,cn=config' nsDS5ReplicaHost nsds5replicaLastUpdateStatus
++ awk '{if ($0 ~ /nsDS5ReplicaHost/) printf("%s",$2); if ($0 ~ /nsds5replicaLastUpdateStatus/) printf(" %s\n",$2)}'
++ return 0
+ agreements='freeipa2 18'
+ for agreement in $agreements
+ printf '%s\n' 'freeipa2 18'
++ return 0
+ agreements='freeipa1 0'
+ for agreement in $agreements
+ printf '%s\n' 'freeipa1 0'
+ for server in "${SERVERS[@]}"
+ [[ -r /tmp/user/1004/tmp.vDKeVHg4LH/replication/freeipa1.example.org ]]
++ wc -l
+ n=1
+ [[ 1 -gt 0 ]]
+ max=1
+ for server in "${SERVERS[@]}"
+ [[ -r /tmp/user/1004/tmp.vDKeVHg4LH/replication/freeipa2.example.org ]]
++ wc -l
+ n=1
+ [[ 1 -gt 1 ]]
+ (( i=1 ))
+ (( i<=max ))
+ [[ 1 -eq 1 ]]
+ first_column='Replication Status'
+ printf %-20s 'Replication Status'
+ for server in "${SERVERS[@]}"
+ [[ -s /tmp/user/1004/tmp.vDKeVHg4LH/replication/freeipa1.example.org ]]
++ sed -n 1p /tmp/user/1004/tmp.vDKeVHg4LH/replication/freeipa1.example.org
+ line='freeipa2 18'
+ sed -i -e 1d /tmp/user/1004/tmp.vDKeVHg4LH/replication/freeipa1.example.org
+ printf %-12s 'freeipa2 18'
+ [[ freeipa2 18 == \E\R\R\O\R ]]
+ for server in "${SERVERS[@]}"
+ [[ -s /tmp/user/1004/tmp.vDKeVHg4LH/replication/freeipa2.example.org ]]
++ sed -n 1p /tmp/user/1004/tmp.vDKeVHg4LH/replication/freeipa2.example.org
+ line='freeipa1 0'
+ sed -i -e 1d /tmp/user/1004/tmp.vDKeVHg4LH/replication/freeipa2.example.org
+ printf %-12s 'freeipa1 0'
+ [[ freeipa1 0 == \E\R\R\O\R ]]
+ printf '%-5s\n' ''
+ (( i++ ))
+ (( i<=max ))
+ set +x
```

**Comment by peterpakos** [comment]
Duplicate of #33.

**Comment by jh23453** [comment]
I disagree - it's not a duplicate :-)

The following patch fixes it for me:
```
--- ipa_check_consistency.orig  2017-12-06 21:26:42.822877584 +0100
+++ ipa_check_consistency       2017-12-06 21:25:31.914642498 +0100
@@ -751,6 +751,7 @@
     fi
     [[ $n -gt $max ]] && max=$n
   done
+  state="OK"
   for ((i=1; i<=max; i++)); do
     [[ $i -eq 1 ]] && first_column="Replication Status" || first_column=""
     printf "%-${FIRST_COLUMN}s" "$first_column"
@@ -758,11 +759,12 @@
       if [[ -s "${dir}/${server}" ]]; then
         line=$(sed -n '1p' "${dir}/${server}")
         sed -i -e '1d' "${dir}/${server}"
+        replstate=$(echo $line | cut -f2 -d " ")
       else
         line=""
       fi
       printf "%-${MIDDLE_COLUMNS}s" "$line"
-      if [[ "$line" == "ERROR" ]]; then
+      if [[ "$replstate" != "0" ]]; then
         state="FAIL"
       fi
     done
```
Output without patch:
```
Anonymous BIND      YES         YES         OK   
Replication Status  freeipa2 18 freeipa1 0  
```
Output with patch and replication error:
```
Anonymous BIND      YES         YES         OK   
Replication Status  freeipa2 18 freeipa1 0  FAIL
```
Output with patch and without replication error:
```
Anonymous BIND      YES         YES         OK   
Replication Status  freeipa2 0  freeipa1 0  OK  
```

**Comment by peterpakos** [comment]
This patch is incorrect as it doesn't take into account multiple replica agreements which will span multiple lines:
```
Replication Status  ipa04 0  ipa04 0  ipa04 0  ipa03 0
                    ipa03 0  ipa03 0  ipa02 0  ipa02 0
                    ipa02 0  ipa01 0  ipa02 0  ipa01 0
```

Plus not all non-zero status codes indicate a failure (see [link](https://access.redhat.com/documentation/en-us/red_hat_directory_server/10/html/configuration_command_and_file_reference/replication_agreement_status) for more info).

Also, your snippet is not in-line with the rest of the script - you won't find a single `echo` in the code.

**Comment by jh23453** [comment]
I'll check the status codes and update my pull request.

**Comment by jh23453** [comment]
Concerning the status codes: I've read the link you provided (Thanks!).
First, status codes are reused for different stages and error conditions, so we might need
to store more info from LDAP during the checks if we really want to analyze deeper.
.
On the other hand, most status codes signal error conditions. Codes for messages like "Backing off, will retry update later." or "Will retry later." designate temporary conditions - but at least my state
looked stable for me (but I didn't wait too long). So I would've been happy with being alerted.

In short: for my usecase I'd be happy with "!= 0" for alerts and I'm unsure what states we might
consider OK too. Do you have suggestions what codes we should consider OK?

For now I'd extract the check into a function which we might expand/change later.

**Comment by peterpakos** [comment]
@jh23453, I've just pushed commit 85474cc to branch issue_34.

Please let me know if it works as expected for you.

**Comment by peterpakos** [comment]
@jh23453, I've just pushed commit c076272 to branch issue_34.

Please let me know if it works as expected for you.

**Comment by peterpakos** [comment]
@jh23453, I've just pushed a proposed fix to branch issue_34.

Please let me know if it works as expected for you before I push it to the master branch.

**Comment by jh23453** [comment]
I've looked at the code and like it. And the branch works for me. Thanks!

**Comment by peterpakos** [comment]
Great, many thanks @jh23453.

I've now pushed the change to master.

**Comment by peterpakos** [comment]
Fixed by bfb91b3.
